src/kernel.fs

```diff
@@ -0,0 +1,124 @@
+( A SUBROUTINE-THREADED KERNEL FOR DMG-FORTH )
+
+require ./asm.fs
+
+also gb-assembler
+
+
+( TEMPORARY HACK: Don't break the game! )
+$4400 ==>
+
+\ : PS $FFFD -> $FF6F ;  \ Parameter Stack address
+\ : PSP C ;     \ Parameter Stack Pointer
+\ : (PSP) [C] ; \ TOS content
+
+: ps-clear,
+  $EF # C ld, ; ( init PSP to 0 )
+
+( 8c/6b )
+: ps-push-hl,
+  C dec,        ( 1c/1b )
+  H A ld,       ( 1c/1b )
+  A [C] ld,     ( 2c/1b )
+  C dec,        ( 1c/1b )
+  L A ld,       ( 1c/1b )
+  A [C] ld, ;   ( 2c/1b )
+
+( 11c/9b )
+: ps-push-lit,
+  ( nn ) # HL ld, ( 3c/3b )
+  ps-push-hl,   ( 8c/6b )
+;
+
+( 8c/6b )
+: ps-pop-hl,
+  [C] A ld,     ( 2c/1b )
+  A L ld,       ( 1c/1b )
+  C inc,        ( 1c/1b )
+  [C] A ld,     ( 2c/1b )
+  A H ld,       ( 1c/1b )
+  C inc, ;      ( 1c/1b )
+
+( 8c/6b )
+: ps-pop-de,
+  [C] A ld,     ( 2c/1b )
+  A E ld,       ( 1c/1b )
+  C inc,        ( 1c/1b )
+  [C] A ld,     ( 2c/1b )
+  A D ld,       ( 1c/1b )
+  C inc, ;      ( 1c/1b )
+
+( Assume you have the following code
+
+  : double dup + ;
+  : quadruple double double ;
+
+  This kernel uses subroutine-threading Forth. The picture below
+  illustrates how the compiled words would look like:
+
+              +------
+              |  ...                            DUP [asm]
+              +------
+                ^
+                |
+               /
+       +----------+--------+------+
+       | CALL dup | CALL + | exit |             DOUBLE
```

Comment:
that `exit` should be `RET`.

Or, should we also document the tail-call optimization here?

Reply:
ignore this, let's do it when TCO is implemented.